packages/gitbook/src/components/DocumentView/CodeBlock/ClientCodeBlock.tsx

```diff
@@ -0,0 +1,26 @@
+'use client';
+
+import type { DocumentBlockCode } from '@gitbook/api';
+import { useEffect, useState } from 'react';
+
+import type { HighlightLine, RenderedInline } from './highlight';
+import type { BlockProps } from '../Block';
+import { ClientCodeBlockRenderer } from './CodeBlockRenderer';
+import { plainHighlight } from './plain-highlight';
+
+type ClientBlockProps = Pick<BlockProps<DocumentBlockCode>, 'block' | 'style'> & {
+    inlines: RenderedInline[];
+};
+
+/**
+ * Render a code-block client-side by calling a server actions to highlight the code.
+ * It allows us to defer some load to avoid blocking the rendering of the whole page with block highlighting.
+ */
+export function ClientCodeBlock(props: ClientBlockProps) {
+    const { block, style, inlines } = props;
+    const [lines, setLines] = useState<HighlightLine[]>(() => plainHighlight(block, []));
+    useEffect(() => {
+        import('./highlight').then(({ highlight }) => highlight(block, inlines).then(setLines));
```

Comment:
Right now we do this at startup time for all code blocks, it can potentially slow down the loading especially on mobile (as it takes CPU).

What do you think of moving this to use an IntersectionObserver and only do it when it gets close to the screen viewport?

I was thinking about this for the server action approach as well.

Reply:
> What do you think of moving this to use an IntersectionObserver and only do it when it gets close to the screen viewport?

Yes I was thinking about this. I will do it now.